src/shutdown.rs

```diff
@@ -265,24 +269,26 @@ impl SourceShutdownCoordinator {
     }
 
     /// Returned future will finish once all sources have finished.
-    pub fn shutdown_tripwire(&self) -> impl Future<Item = (), Error = ()> {
-        future::join_all(
-            self.shutdown_complete_tripwires
-                .values()
-                .cloned()
-                .collect::<Vec<_>>(),
-        )
-        .map(|_| info!("All sources have finished."))
+    pub fn shutdown_tripwire(&self) -> future::BoxFuture<'static, ()> {
+        let mut futures = Vec::with_capacity(self.shutdown_complete_tripwires.len());
```

Comment:
`collect` is smart enough to do this optimization.

Reply:
I had lifetime issues with `collect`, will try again :+1:

Reply:
Solved issue by adding `.boxed()` to iterator.